feat: add accessor array support to `blas/base/gscal`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add accessor array support including relevant tests and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/24

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add accessor array support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/lib/accessors.js:
```diff
@@ -0,0 +1,102 @@
+
+	ix = offset;
+	if ( stride === 0 ) {
+		x[ 0 ] *= N * alpha;
```
```suggestion
```

kgryte committed:
fix: remove copy-paste mistake

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/lib/accessors.js:
```diff
@@ -0,0 +1,101 @@
+		return x;
+	}
+	for ( i = 0; i < N; i++ ) {
+		x[ ix ] *= alpha;
```
```suggestion
```

kgryte committed:
fix: remove copy-paste mistake

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/lib/accessors.js:
```diff
@@ -0,0 +1,100 @@
+			return x;
+		}
+		for ( i = m; i < N; i += M ) {
+			xset( xbuf, i, xget( xbuf, i ) * alpha );
+			xset( xbuf, i+1, xget( xbuf, i+1 ) * alpha );
+			xset( xbuf, i+2, xget( xbuf, i+2 ) * alpha );
+			xset( xbuf, i+3, xget( xbuf, i+3 ) * alpha );
+			xset( xbuf, i+4, xget( xbuf, i+4 ) * alpha );
```
```suggestion
			xset( xbuf, ix, xget( xbuf, ix ) * alpha );
			xset( xbuf, ix+1, xget( xbuf, ix+1 ) * alpha );
			xset( xbuf, ix+2, xget( xbuf, ix+2 ) * alpha );
			xset( xbuf, ix+3, xget( xbuf, ix+3 ) * alpha );
			xset( xbuf, ix+4, xget( xbuf, ix+4 ) * alpha );
			ix += M;
```

kgryte committed:
fix: use correct offset

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
fix: use correct offset

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/lib/ndarray.js:
```diff
@@ -44,12 +50,23 @@ var M = 5;
+	}
+
 	ix = offset;
+	if ( stride === 0) {
```
```suggestion
	if ( stride === 0 ) {
```

kgryte committed:
style: add missing space

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.main.js:
```diff
@@ -139,17 +188,99 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	]);
+
+	gscal( 3, 5.0, x, 2 );
+	t.deepEqual( x, expected, 'returns expected value' );
```
I believe we've discussed this before, but this does not behave as you think it does. `t.deepEqual` is not accessor aware.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.main.js:
```diff
@@ -139,17 +188,99 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	var expected;
+	var x;
+
+	x = toAccessorArray([
```
Same thing.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.main.js:
```diff
@@ -139,17 +188,99 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	expected = toAccessorArray( [ 45.0, -4.0, 1.0 ] );
+
+	gscal( x.length, 5.0, x, 0 );
 	t.deepEqual( x, expected, 'returns expected value' );
```
Again, this doesn't do what you think it does.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.ndarray.js:
```diff
@@ -75,6 +76,43 @@ tape( 'the function multiplies `x` by a constant', function test( t ) {
+	]);
+
+	gscal( x.length, 5.0, x, 1, 0 );
+	t.deepEqual( x, expected, 'returns expected value' );
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.ndarray.js:
```diff
@@ -139,6 +214,30 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	var expected;
+	var x;
+
+	x = toAccessorArray([
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.ndarray.js:
```diff
@@ -163,6 +262,30 @@ tape( 'the function supports specifying a negative stride', function test( t ) {
+	var expected;
+	var x;
+
+	x = toAccessorArray([
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.ndarray.js:
```diff
@@ -189,6 +312,32 @@ tape( 'the function supports an offset parameter', function test( t ) {
+	var expected;
+	var x;
+
+	x = toAccessorArray([
```
Same comments.

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.main.js:
```diff
@@ -139,17 +188,99 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	var expected;
+	var x;
+
+	x = [
+		2.0,  // 0
+		-3.0,
+		-5.0, // 1
+		7.0,
+		6.0   // 2
+	];
+	expected = [
+		10.0,  // 0
+		-3.0,
+		-25.0, // 1
+		7.0,
+		30.0   // 2
+	];
```
```suggestion
	x = [
		2.0,  // 2
		-3.0,
		-5.0, // 1
		7.0,
		6.0   // 0
	];
	expected = [
		10.0,  // 2
		-3.0,
		-25.0, // 1
		7.0,
		30.0   // 0
	];
```

kgryte replied:
Indices should be numbered in iteration order.

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.main.js:
```diff
@@ -139,17 +188,99 @@ tape( 'the function supports specifying a stride', function test( t ) {
+	var x;
+
+	x = [
+		2.0,  // 0
+		-3.0,
+		-5.0, // 1
+		7.0,
+		6.0   // 2
+	];
+	expected = [
+		10.0,  // 0
+		-3.0,
+		-25.0, // 1
+		7.0,
+		30.0   // 2
```
```suggestion
		2.0,  // 2
		-3.0,
		-5.0, // 1
		7.0,
		6.0   // 0
	];
	expected = [
		10.0,  // 2
		-3.0,
		-25.0, // 1
		7.0,
		30.0   // 0
```

kgryte commented on lib/node_modules/@stdlib/blas/base/gscal/test/test.ndarray.js:
```diff
@@ -163,6 +262,30 @@ tape( 'the function supports specifying a negative stride', function test( t ) {
+		6.0   // 0
+	];
+	expected = [
+		10.0,  // 0
+		-3.0,
+		-25.0, // 1
+		7.0,
+		30.0   // 2
```
```suggestion
		10.0,  // 2
		-3.0,
		-25.0, // 1
		7.0,
		30.0   // 0
```

headlessNode committed:
chore: apply code review suggestion

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
chore: apply code review suggestion

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
chore: apply code review suggestion

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte committed:
docs: add comments

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
